File: random-flip-matrix/random-flip-matrix.java
class Solution {
    
    private int rows, cols, total;
    private Random rand;
    private Set<Integer> available; // this contains a set of coordinates that we could set to one

    public Solution(int n_rows, int n_cols) {
        this.rows = n_rows;
        this.cols = n_cols;
        this.rand = new Random();
        this.available = new HashSet<>();
        this.total = this.rows * this.cols;
    }
    
    public int[] flip() {
        int x = rand.nextInt(this.total);
        
        // keep grabbing random numbers until we find one that isn't in there
        while (available.contains(x)) {
            x = rand.nextInt(this.total);
        }
        
        this.available.add(x);
        return new int[]{x / this.cols, x % this.cols};
    }
    
    public void reset() {
        this.available.clear();
    }
}

/**
 * Your Solution object will be instantiated and called as such:
 * Solution obj = new Solution(n_rows, n_cols);
 * int[] param_1 = obj.flip();
 * obj.reset();
 */

File: 913-random-flip-matrix/random-flip-matrix.py
class Solution:
    def __init__(self, m: int, n: int):
        self.n, self.len = n, m * n
        self.items = {}
        self.last_idx = self.len-1

    def flip(self) -> List[int]:
        idx = random.randint(0, self.last_idx)
        v = self.items.get(idx, idx)
        self.items[idx] = self.items.get(self.last_idx, self.last_idx)
        self.last_idx -= 1
        return divmod(v, self.n)

    def reset(self) -> None:
        self.items.clear()
        self.last_idx = self.len-1

# Your Solution object will be instantiated and called as such:
# obj = Solution(m, n)
# param_1 = obj.flip()
# obj.reset()